src/pydocstyle/parser.py

```diff
@@ -121,7 +122,8 @@ class Module(Definition):
 
     @property
     def is_public(self):
-        return not self.name.startswith('_') or self.name.startswith('__')
+        module_name, _ = os.path.splitext(os.path.basename(self.name))
```

Comment:
Now that I see this and the tests, I think the fix needs to be a little more complex, see [How publicity is determined](http://www.pydocstyle.org/en/2.1.1/error_codes.html#how-publicity-is-determined).

We need to look at _all_ the packages in this module's path to determine the publicity, something like
```.py
def is_public(self):
    def is_public_name(module_name):
        return not module_name.startswith('_') or module_name.startswith('__')
    return all(is_public_name(name) for name in self.name.split('.'))
```
Also, I would appreciate a comment describing the logic here.

Reply:
I like that except `Module.name` is actually the filepath. It's a bit confusing though so I think `path` or `filepath` should be a new attribute on `name` should be what `module.__name__` evaluates to in python.

The tricky part is getting the module name. Particularly from absolute paths or relative ones run from within a package. Like if run pydocstyle in `mylib/_subpackage/` or give it relative filepath, it won't see the `_subpackage` part right now.

Can you confirm whether or not the following should all work and are correct?

`file.py` module name is `package._subpackage.file` and is private:
* `~ $ pydocstyle ~/project/package/_subpackage/file.py`
* `~/project $ pydocstyle ~/project/package/_subpackage/file.py`
* `~/project/package/_subpackage $ pydocstyle file.py`

`__init__.py` module name is `package` and is public:
* `~ $ pydocstyle _project/package/__init__.py`
* `~/_project $ pydocstyle package/__init__.py`

I don't actually know how that would be implemented, I don't even want to think about namespace packages.

Reply:
Those are some good points. I'm not sure. Perhaps we need to keep this as you implemented and document that the publicity logic only applies up to the module level.

@shacharoo, do you want to weigh in?